From ddbdd606ae7ead7f10e0f1a47ebfd8b13a682df6 Mon Sep 17 00:00:00 2001
From: JS Deck <jsdeckerido@gmail.com>
Date: Mon, 11 May 2026 18:23:58 -0300
Subject: [PATCH 02/16] drm/msm/dp: use common msm_dp_mode between panel and
 display; introduce msm_dp_panel_mode_cfg to indicate mode params including
 bpp, dsc, and fec state; update msm_dp_panel_get_supported_bpp to
 msm_dp_panel_get_supported_cfg to report bpp, dsc, and fec state; fix
 mode_pclk_khz being used with link rate comparison

---
 drivers/gpu/drm/msm/dp/dp_ctrl.c    |   7 +-
 drivers/gpu/drm/msm/dp/dp_debug.c   |   6 +-
 drivers/gpu/drm/msm/dp/dp_display.c |  65 ++++------
 drivers/gpu/drm/msm/dp/dp_panel.c   | 180 +++++++++++++++++++++++-----
 drivers/gpu/drm/msm/dp/dp_panel.h   |  17 ++-
 drivers/gpu/drm/msm/msm_drv.h       |   7 ++
 6 files changed, 207 insertions(+), 75 deletions(-)

diff --git a/drivers/gpu/drm/msm/dp/dp_ctrl.c b/drivers/gpu/drm/msm/dp/dp_ctrl.c
index ef298c7d3e5e..a69ba74f5153 100644
--- a/drivers/gpu/drm/msm/dp/dp_ctrl.c
+++ b/drivers/gpu/drm/msm/dp/dp_ctrl.c
@@ -404,7 +404,7 @@ static void msm_dp_ctrl_config_ctrl(struct msm_dp_ctrl_private *ctrl)
 		config |= DP_CONFIGURATION_CTRL_ASSR;
 
 	tbd = msm_dp_link_get_test_bits_depth(ctrl->link,
-			ctrl->panel->msm_dp_mode.bpp);
+			ctrl->panel->msm_dp_mode.mode_cfg.bpp);
 
 	config |= tbd << DP_CONFIGURATION_CTRL_BPC_SHIFT;
 
@@ -452,7 +452,8 @@ static void msm_dp_ctrl_configure_source_params(struct msm_dp_ctrl_private *ctrl
 
 	msm_dp_ctrl_config_ctrl(ctrl);
 
-	test_bits_depth = msm_dp_link_get_test_bits_depth(ctrl->link, ctrl->panel->msm_dp_mode.bpp);
+	test_bits_depth = msm_dp_link_get_test_bits_depth(ctrl->link,
+				ctrl->panel->msm_dp_mode.mode_cfg.bpp);
 	colorimetry_cfg = msm_dp_link_get_colorimetry_config(ctrl->link);
 
 	misc_val = msm_dp_read_link(ctrl, REG_DP_MISC1_MISC0);
@@ -1249,7 +1250,7 @@ static void msm_dp_ctrl_calc_tu_parameters(struct msm_dp_ctrl_private *ctrl,
 	in.hactive = drm_mode->hdisplay;
 	in.hporch = drm_mode->htotal - drm_mode->hdisplay;
 	in.nlanes = ctrl->link->link_params.num_lanes;
-	in.bpp = ctrl->panel->msm_dp_mode.bpp;
+	in.bpp = ctrl->panel->msm_dp_mode.mode_cfg.bpp;
 	in.pixel_enc = ctrl->panel->msm_dp_mode.out_fmt_is_yuv_420 ? 420 : 444;
 	in.dsc_en = 0;
 	in.async_en = 0;
diff --git a/drivers/gpu/drm/msm/dp/dp_debug.c b/drivers/gpu/drm/msm/dp/dp_debug.c
index cf3838fcd154..18fc3b6f7902 100644
--- a/drivers/gpu/drm/msm/dp/dp_debug.c
+++ b/drivers/gpu/drm/msm/dp/dp_debug.c
@@ -65,7 +65,11 @@ static int msm_dp_debug_show(struct seq_file *seq, void *p)
 	seq_printf(seq, "\t\tpixel clock khz = %d\n",
 			drm_mode->clock);
 	seq_printf(seq, "\t\tbpp = %d\n",
-			debug->panel->msm_dp_mode.bpp);
+			debug->panel->msm_dp_mode.mode_cfg.bpp);
+	seq_printf(seq, "\t\tfec = %d\n",
+			debug->panel->msm_dp_mode.mode_cfg.fec_available);
+	seq_printf(seq, "\t\tdsc = %d\n",
+			debug->panel->msm_dp_mode.mode_cfg.dsc);
 
 	/* Link Information */
 	seq_printf(seq, "\tdp_link:\n\t\ttest_requested = %d\n",
diff --git a/drivers/gpu/drm/msm/dp/dp_display.c b/drivers/gpu/drm/msm/dp/dp_display.c
index d2124d625485..c18e6b1cb979 100644
--- a/drivers/gpu/drm/msm/dp/dp_display.c
+++ b/drivers/gpu/drm/msm/dp/dp_display.c
@@ -91,7 +91,6 @@ struct msm_dp_display_private {
 	struct msm_dp_panel   *panel;
 	struct msm_dp_ctrl    *ctrl;
 
-	struct msm_dp_display_mode msm_dp_mode;
 	struct msm_dp msm_dp_display;
 
 	/* wait for audio signaling */
@@ -828,20 +827,6 @@ static int msm_dp_init_sub_modules(struct msm_dp_display_private *dp)
 	return rc;
 }
 
-static int msm_dp_display_set_mode(struct msm_dp *msm_dp_display,
-			       struct msm_dp_display_mode *mode)
-{
-	struct msm_dp_display_private *dp;
-
-	dp = container_of(msm_dp_display, struct msm_dp_display_private, msm_dp_display);
-
-	drm_mode_copy(&dp->panel->msm_dp_mode.drm_mode, &mode->drm_mode);
-	dp->panel->msm_dp_mode.bpp = mode->bpp;
-	dp->panel->msm_dp_mode.out_fmt_is_yuv_420 = mode->out_fmt_is_yuv_420;
-	msm_dp_panel_init_panel_info(dp->panel);
-	return 0;
-}
-
 static int msm_dp_display_enable(struct msm_dp_display_private *dp, bool force_link_train)
 {
 	int rc = 0;
@@ -935,10 +920,10 @@ enum drm_mode_status msm_dp_bridge_mode_valid(struct drm_bridge *bridge,
 {
 	const u32 num_components = 3, default_bpp = 24;
 	struct msm_dp_display_private *msm_dp_display;
-	struct msm_dp_link_info *link_info;
-	u32 mode_rate_khz = 0, supported_rate_khz = 0, mode_bpp = 0;
+	u32 mode_bpp;
 	struct msm_dp *dp;
 	int mode_pclk_khz = mode->clock;
+	struct msm_dp_display_mode_cfg mode_cfg;
 
 	dp = to_dp_bridge(bridge)->msm_dp_display;
 
@@ -948,7 +933,6 @@ enum drm_mode_status msm_dp_bridge_mode_valid(struct drm_bridge *bridge,
 	}
 
 	msm_dp_display = container_of(dp, struct msm_dp_display_private, msm_dp_display);
-	link_info = &msm_dp_display->panel->link_info;
 
 	if ((drm_mode_is_420_only(&dp->connector->display_info, mode) &&
 	     msm_dp_display->panel->vsc_sdp_supported) ||
@@ -962,13 +946,14 @@ enum drm_mode_status msm_dp_bridge_mode_valid(struct drm_bridge *bridge,
 	if (!mode_bpp)
 		mode_bpp = default_bpp;
 
-	mode_bpp = msm_dp_panel_get_mode_bpp(msm_dp_display->panel,
-			mode_bpp, mode_pclk_khz);
-
-	mode_rate_khz = mode_pclk_khz * mode_bpp;
-	supported_rate_khz = link_info->num_lanes * link_info->rate * 8;
+	/* Need to use clock not-scaled by yuv/wide-bus for link rate check */
+	mode_cfg = msm_dp_panel_get_mode_cfg(msm_dp_display->panel,
+			mode_bpp,
+			msm_dp_display->panel->dsc_cap.supported ?
+					MSM_MODE_DSC_OPTIONAL : MSM_MODE_DSC_UNAVAILABLE,
+			mode->clock);
 
-	if (mode_rate_khz > supported_rate_khz)
+	if (mode_cfg.bpp == MSM_DP_DISPLAY_MODE_BPP_UNAVAILABLE)
 		return MODE_BAD;
 
 	return MODE_OK;
@@ -1539,7 +1524,7 @@ bool msm_dp_wide_bus_available(const struct msm_dp *msm_dp_display)
 
 	dp = container_of(msm_dp_display, struct msm_dp_display_private, msm_dp_display);
 
-	if (dp->msm_dp_mode.out_fmt_is_yuv_420)
+	if (dp->panel->msm_dp_mode.out_fmt_is_yuv_420)
 		return false;
 
 	return dp->wide_bus_supported;
@@ -1601,7 +1586,7 @@ void msm_dp_bridge_atomic_enable(struct drm_bridge *drm_bridge,
 	bool force_link_train = false;
 
 	msm_dp_display = container_of(dp, struct msm_dp_display_private, msm_dp_display);
-	if (!msm_dp_display->msm_dp_mode.drm_mode.clock) {
+	if (!msm_dp_display->panel->msm_dp_mode.drm_mode.clock) {
 		DRM_ERROR("invalid params\n");
 		return;
 	}
@@ -1622,7 +1607,7 @@ void msm_dp_bridge_atomic_enable(struct drm_bridge *drm_bridge,
 		return;
 	}
 
-	rc = msm_dp_display_set_mode(dp, &msm_dp_display->msm_dp_mode);
+	rc = msm_dp_panel_init_panel_info(msm_dp_display->panel);
 	if (rc) {
 		DRM_ERROR("Failed to perform a mode set, rc=%d\n", rc);
 		mutex_unlock(&msm_dp_display->event_mutex);
@@ -1707,35 +1692,37 @@ void msm_dp_bridge_mode_set(struct drm_bridge *drm_bridge,
 	struct msm_dp *dp = msm_dp_bridge->msm_dp_display;
 	struct msm_dp_display_private *msm_dp_display;
 	struct msm_dp_panel *msm_dp_panel;
+	struct msm_dp_display_mode *msm_dp_mode;
 
 	msm_dp_display = container_of(dp, struct msm_dp_display_private, msm_dp_display);
 	msm_dp_panel = msm_dp_display->panel;
+	msm_dp_mode = &msm_dp_display->panel->msm_dp_mode;
 
-	memset(&msm_dp_display->msm_dp_mode, 0x0, sizeof(struct msm_dp_display_mode));
+	memset(msm_dp_mode, 0x0, sizeof(struct msm_dp_display_mode));
 
 	if (msm_dp_display_check_video_test(dp))
-		msm_dp_display->msm_dp_mode.bpp = msm_dp_display_get_test_bpp(dp);
+		msm_dp_mode->mode_cfg.bpp = msm_dp_display_get_test_bpp(dp);
 	else /* Default num_components per pixel = 3 */
-		msm_dp_display->msm_dp_mode.bpp = dp->connector->display_info.bpc * 3;
+		msm_dp_mode->mode_cfg.bpp = dp->connector->display_info.bpc * 3;
 
-	if (!msm_dp_display->msm_dp_mode.bpp)
-		msm_dp_display->msm_dp_mode.bpp = 24; /* Default bpp */
+	if (!msm_dp_mode->mode_cfg.bpp)
+		msm_dp_mode->mode_cfg.bpp = 24; /* Default bpp */
 
-	drm_mode_copy(&msm_dp_display->msm_dp_mode.drm_mode, adjusted_mode);
+	drm_mode_copy(&msm_dp_mode->drm_mode, adjusted_mode);
 
-	msm_dp_display->msm_dp_mode.v_active_low =
-		!!(msm_dp_display->msm_dp_mode.drm_mode.flags & DRM_MODE_FLAG_NVSYNC);
+	msm_dp_mode->v_active_low =
+		!!(msm_dp_mode->drm_mode.flags & DRM_MODE_FLAG_NVSYNC);
 
-	msm_dp_display->msm_dp_mode.h_active_low =
-		!!(msm_dp_display->msm_dp_mode.drm_mode.flags & DRM_MODE_FLAG_NHSYNC);
+	msm_dp_mode->h_active_low =
+		!!(msm_dp_mode->drm_mode.flags & DRM_MODE_FLAG_NHSYNC);
 
-	msm_dp_display->msm_dp_mode.out_fmt_is_yuv_420 =
+	msm_dp_mode->out_fmt_is_yuv_420 =
 		drm_mode_is_420_only(&dp->connector->display_info, adjusted_mode) &&
 		msm_dp_panel->vsc_sdp_supported;
 
 	/* populate wide_bus_support to different layers */
 	msm_dp_display->ctrl->wide_bus_en =
-		msm_dp_display->msm_dp_mode.out_fmt_is_yuv_420 ? false : msm_dp_display->wide_bus_supported;
+		msm_dp_mode->out_fmt_is_yuv_420 ? false : msm_dp_display->wide_bus_supported;
 }
 
 void msm_dp_bridge_hpd_enable(struct drm_bridge *bridge)
diff --git a/drivers/gpu/drm/msm/dp/dp_panel.c b/drivers/gpu/drm/msm/dp/dp_panel.c
index ba0c93d9ac5b..1a551e9ff247 100644
--- a/drivers/gpu/drm/msm/dp/dp_panel.c
+++ b/drivers/gpu/drm/msm/dp/dp_panel.c
@@ -11,6 +11,7 @@
 #include <drm/drm_edid.h>
 #include <drm/drm_of.h>
 #include <drm/drm_print.h>
+#include <drm/drm_fixed.h>
 
 #include <linux/io.h>
 #include <linux/types.h>
@@ -211,25 +212,136 @@ static int msm_dp_panel_read_dpcd(struct msm_dp_panel *msm_dp_panel)
 	return rc;
 }
 
-static u32 msm_dp_panel_get_supported_bpp(struct msm_dp_panel *msm_dp_panel,
-		u32 mode_edid_bpp, u32 mode_pclk_khz)
+static u32 msm_dp_panel_calc_link_rate(struct msm_dp_panel *msm_dp_panel,
+			bool fec_en)
 {
 	const struct msm_dp_link_info *link_info;
-	const u32 max_supported_bpp = 30, min_supported_bpp = 18;
-	u32 bpp, data_rate_khz;
-
-	bpp = min(mode_edid_bpp, max_supported_bpp);
+	s64 rate_fp;
+	s64 fec_overhead_fp;
+	u32 data_rate_khz;
 
 	link_info = &msm_dp_panel->link_info;
 	data_rate_khz = link_info->num_lanes * link_info->rate * 8;
 
-	do {
+	if (fec_en) {
+		fec_overhead_fp = drm_fixp_from_fraction(100000, 97582);
+
+		rate_fp = drm_int2fixp(data_rate_khz);
+		rate_fp = drm_fixp_div(rate_fp, fec_overhead_fp);
+		data_rate_khz = drm_fixp2int(rate_fp);
+	}
+
+	return data_rate_khz;
+}
+
+static u32 msm_dp_panel_get_supported_bpp_no_dsc(
+		u32 mode_edid_bpp,
+		u32 mode_pclk_khz,
+		u32 data_rate_khz)
+{
+	const u32 max_supported_bpp = 30, min_supported_bpp = 18;
+	u32 bpp = min(mode_edid_bpp, max_supported_bpp);
+
+	for (; bpp >= min_supported_bpp; bpp -= 6)
 		if (mode_pclk_khz * bpp <= data_rate_khz)
 			return bpp;
-		bpp -= 6;
-	} while (bpp > min_supported_bpp);
 
-	return min_supported_bpp;
+	return MSM_DP_DISPLAY_MODE_BPP_UNAVAILABLE;
+}
+
+#define MSM_DP_MIN_SUPPORTED_DSC_BPP 24
+static u32 msm_dp_panel_get_supported_bpp_dsc(
+		u32 mode_edid_bpp,
+		u32 mode_pclk_khz,
+		u32 data_rate_khz,
+		u8 bpc[3])
+{
+	const u32 max_supported_bpp = min(mode_edid_bpp, 30);
+	const u32 min_supported_bpp = MSM_DP_MIN_SUPPORTED_DSC_BPP;
+	const u32 num_components = 3;
+	u32 max_bpp = MSM_DP_DISPLAY_MODE_BPP_UNAVAILABLE;
+	int i;
+
+	for (i = 0; i < 3; i++) {
+		if (bpc[i]) {
+			u32 bpp = bpc[i] * num_components;
+
+			if (bpp > max_bpp &&
+						bpp >= min_supported_bpp &&
+						bpp <= max_supported_bpp &&
+						mode_pclk_khz * bpp / MSM_DP_DSC_COMP_RATIO <=
+									data_rate_khz)
+				max_bpp = bpp;
+		}
+	}
+
+	return max_bpp;
+}
+
+static struct msm_dp_display_mode_cfg msm_dp_panel_get_cfg_from_bpp(
+	u32 bpp, u32 dsc_bpp, enum msm_mode_dsc_cfg dsc_cfg, bool fec_en)
+{
+	const u32 min_supported_bpp = MSM_DP_MIN_SUPPORTED_DSC_BPP;
+	struct msm_dp_display_mode_cfg cfg = {0};
+
+	cfg.fec_available = fec_en;
+
+	if (!dsc_bpp)
+		dsc_cfg = MSM_MODE_DSC_UNAVAILABLE;
+	else if (!bpp)
+		dsc_cfg = MSM_MODE_DSC_REQUIRED;
+
+	switch (dsc_cfg) {
+	default:
+	case MSM_MODE_DSC_UNAVAILABLE:
+		cfg.bpp = bpp;
+		break;
+	case MSM_MODE_DSC_OPTIONAL:
+	case MSM_MODE_DSC_PREFERRED:
+		/*
+		 * Return "if dsc is preferred" when not requested to be required.
+		 * dsc is preferred if non-dsc bpp is of lower bpp than
+		 * min supported dsc bpp
+		 */
+		dsc_cfg = bpp < min_supported_bpp ?
+					MSM_MODE_DSC_PREFERRED :
+					MSM_MODE_DSC_OPTIONAL;
+		cfg.bpp = bpp;
+		break;
+	case MSM_MODE_DSC_REQUIRED:
+		cfg.bpp = dsc_bpp;
+		break;
+	}
+
+	cfg.dsc = dsc_cfg;
+
+	return cfg;
+}
+
+static struct msm_dp_display_mode_cfg msm_dp_panel_get_supported_cfg(
+		struct msm_dp_panel *msm_dp_panel,
+		u32 mode_edid_bpp,
+		enum msm_mode_dsc_cfg dsc_cfg,
+		u32 mode_pclk_khz)
+{
+	u32 bpp = MSM_DP_DISPLAY_MODE_BPP_UNAVAILABLE;
+	u32 dsc_bpp = MSM_DP_DISPLAY_MODE_BPP_UNAVAILABLE;
+	u32 data_rate_khz;
+	bool fec_en;
+
+	fec_en = msm_dp_panel->fec_cap.supported;
+	data_rate_khz = msm_dp_panel_calc_link_rate(msm_dp_panel, fec_en);
+
+	if (dsc_cfg < MSM_MODE_DSC_REQUIRED)
+		bpp = msm_dp_panel_get_supported_bpp_no_dsc(
+					mode_edid_bpp, mode_pclk_khz, data_rate_khz);
+
+	if (dsc_cfg > MSM_MODE_DSC_UNAVAILABLE)
+		dsc_bpp = msm_dp_panel_get_supported_bpp_dsc(
+					mode_edid_bpp, mode_pclk_khz, data_rate_khz,
+					msm_dp_panel->dsc_cap.bpc);
+
+	return msm_dp_panel_get_cfg_from_bpp(bpp, dsc_bpp, dsc_cfg, fec_en);
 }
 
 static void msm_dp_panel_read_sink_fec_caps(struct msm_dp_panel_private *panel)
@@ -367,27 +479,30 @@ int msm_dp_panel_read_sink_caps(struct msm_dp_panel *msm_dp_panel,
 	return rc;
 }
 
-u32 msm_dp_panel_get_mode_bpp(struct msm_dp_panel *msm_dp_panel,
-		u32 mode_edid_bpp, u32 mode_pclk_khz)
+struct msm_dp_display_mode_cfg msm_dp_panel_get_mode_cfg(
+		struct msm_dp_panel *msm_dp_panel,
+		u32 mode_edid_bpp,
+		enum msm_mode_dsc_cfg dsc_cfg,
+		u32 mode_pclk_khz)
 {
 	struct msm_dp_panel_private *panel;
-	u32 bpp;
+	struct msm_dp_display_mode_cfg cfg = {0};
 
 	if (!msm_dp_panel || !mode_edid_bpp || !mode_pclk_khz) {
 		DRM_ERROR("invalid input\n");
-		return 0;
+		return cfg;
 	}
 
 	panel = container_of(msm_dp_panel, struct msm_dp_panel_private, msm_dp_panel);
 
-	if (msm_dp_panel->video_test)
-		bpp = msm_dp_link_bit_depth_to_bpp(
+	if (msm_dp_panel->video_test) {
+		cfg.bpp = msm_dp_link_bit_depth_to_bpp(
 				panel->link->test_video.test_bit_depth);
-	else
-		bpp = msm_dp_panel_get_supported_bpp(msm_dp_panel, mode_edid_bpp,
-				mode_pclk_khz);
-
-	return bpp;
+		return cfg;
+	} else {
+		return msm_dp_panel_get_supported_cfg(msm_dp_panel, mode_edid_bpp,
+				dsc_cfg, mode_pclk_khz);
+	}
 }
 
 int msm_dp_panel_get_modes(struct msm_dp_panel *msm_dp_panel,
@@ -659,7 +774,7 @@ static int msm_dp_panel_setup_vsc_sdp_yuv_420(struct msm_dp_panel *msm_dp_panel)
 	vsc_sdp_data.colorimetry = DP_COLORIMETRY_DEFAULT;
 
 	/* VSC SDP Payload for DB17 */
-	vsc_sdp_data.bpc = msm_dp_mode->bpp / 3;
+	vsc_sdp_data.bpc = msm_dp_mode->mode_cfg.bpp / 3;
 	vsc_sdp_data.dynamic_range = DP_DYNAMIC_RANGE_CTA;
 
 	/* VSC SDP Payload for DB18 */
@@ -756,9 +871,11 @@ int msm_dp_panel_timing_cfg(struct msm_dp_panel *msm_dp_panel, bool wide_bus_en)
 int msm_dp_panel_init_panel_info(struct msm_dp_panel *msm_dp_panel)
 {
 	struct drm_display_mode *drm_mode;
+	struct msm_dp_display_mode_cfg *mode_cfg;
 	struct msm_dp_panel_private *panel;
 
 	drm_mode = &msm_dp_panel->msm_dp_mode.drm_mode;
+	mode_cfg = &msm_dp_panel->msm_dp_mode.mode_cfg;
 
 	panel = container_of(msm_dp_panel, struct msm_dp_panel_private, msm_dp_panel);
 
@@ -781,13 +898,18 @@ int msm_dp_panel_init_panel_info(struct msm_dp_panel *msm_dp_panel)
 			drm_mode->vsync_end - drm_mode->vsync_start);
 	drm_dbg_dp(panel->drm_dev, "pixel clock (KHz)=(%d)\n",
 				drm_mode->clock);
-	drm_dbg_dp(panel->drm_dev, "bpp = %d\n", msm_dp_panel->msm_dp_mode.bpp);
-
-	msm_dp_panel->msm_dp_mode.bpp = msm_dp_panel_get_mode_bpp(msm_dp_panel, msm_dp_panel->msm_dp_mode.bpp,
-						      msm_dp_panel->msm_dp_mode.drm_mode.clock);
-
-	drm_dbg_dp(panel->drm_dev, "updated bpp = %d\n",
-				msm_dp_panel->msm_dp_mode.bpp);
+	drm_dbg_dp(panel->drm_dev, "bpp = %d\n", mode_cfg->bpp);
+
+	*mode_cfg = msm_dp_panel_get_mode_cfg(
+						      msm_dp_panel,
+						      mode_cfg->bpp,
+						      mode_cfg->dsc,
+						      drm_mode->clock);
+
+	drm_dbg_dp(panel->drm_dev, "updated_bpp=%d fec=%d dsc=%d\n",
+				mode_cfg->bpp,
+				mode_cfg->fec_available,
+				mode_cfg->dsc);
 
 	return 0;
 }
diff --git a/drivers/gpu/drm/msm/dp/dp_panel.h b/drivers/gpu/drm/msm/dp/dp_panel.h
index e036aebece66..5304d3d38c3e 100644
--- a/drivers/gpu/drm/msm/dp/dp_panel.h
+++ b/drivers/gpu/drm/msm/dp/dp_panel.h
@@ -11,12 +11,23 @@
 
 #include "dp_aux.h"
 #include "dp_link.h"
+#include "msm_drv.h"
 
 struct edid;
 
+#define MSM_DP_DSC_COMP_RATIO 3
+
+#define MSM_DP_DISPLAY_MODE_BPP_UNAVAILABLE 0
+
+struct msm_dp_display_mode_cfg {
+	u32 bpp;
+	bool fec_available;
+	enum msm_mode_dsc_cfg dsc;
+};
+
 struct msm_dp_display_mode {
 	struct drm_display_mode drm_mode;
-	u32 bpp;
+	struct msm_dp_display_mode_cfg mode_cfg;
 	u32 h_active_low;
 	u32 v_active_low;
 	bool out_fmt_is_yuv_420;
@@ -65,8 +76,8 @@ int msm_dp_panel_deinit(struct msm_dp_panel *msm_dp_panel);
 int msm_dp_panel_timing_cfg(struct msm_dp_panel *msm_dp_panel, bool wide_bus_en);
 int msm_dp_panel_read_sink_caps(struct msm_dp_panel *msm_dp_panel,
 		struct drm_connector *connector);
-u32 msm_dp_panel_get_mode_bpp(struct msm_dp_panel *msm_dp_panel, u32 mode_max_bpp,
-			u32 mode_pclk_khz);
+struct msm_dp_display_mode_cfg msm_dp_panel_get_mode_cfg(struct msm_dp_panel *msm_dp_panel,
+			u32 mode_max_bpp, enum msm_mode_dsc_cfg dsc_cfg, u32 mode_pclk_khz);
 int msm_dp_panel_get_modes(struct msm_dp_panel *msm_dp_panel,
 		struct drm_connector *connector);
 void msm_dp_panel_handle_sink_request(struct msm_dp_panel *msm_dp_panel);
diff --git a/drivers/gpu/drm/msm/msm_drv.h b/drivers/gpu/drm/msm/msm_drv.h
index 617b3c4b42c0..f54d91835c22 100644
--- a/drivers/gpu/drm/msm/msm_drv.h
+++ b/drivers/gpu/drm/msm/msm_drv.h
@@ -70,6 +70,13 @@ enum msm_dsi_controller {
 
 #define MSM_GPU_MAX_RINGS 4
 
+enum msm_mode_dsc_cfg {
+	MSM_MODE_DSC_UNAVAILABLE,
+	MSM_MODE_DSC_OPTIONAL,
+	MSM_MODE_DSC_PREFERRED,
+	MSM_MODE_DSC_REQUIRED,
+};
+
 struct msm_drm_private {
 
 	struct drm_device *dev;
-- 
2.54.0

